File: caf-smart-auth-react-native-expo/android/src/main/java/io/caf/smartauthbridge/CafSmartAuthBridgeSettings.kt
package io.caf.smartauthbridge

import io.caf.smartauth.input.CafFilterStyle
import io.caf.smartauth.input.CafStage
import org.json.JSONObject
import java.io.Serializable

internal data class CafFaceAuthenticationSettingsModel(
    val loadingScreen: Boolean?,
    val enableScreenCapture: Boolean?,
    val filter: CafFilterStyle?
)

internal class CafSmartAuthBridgeSettings(settings: String) : Serializable {
    val cafStage: CafStage?
    val faceAuthenticatorSettings: CafFaceAuthenticationSettingsModel?

    init {
        val jsonObject = JSONObject(settings)

        cafStage = if (jsonObject.has(STAGE)) CafStage.entries[jsonObject.getInt(STAGE)]
        else null

        faceAuthenticatorSettings = if (jsonObject.has(FACE_AUTHENTICATION_SETTINGS)) {
            val faceAuthenticatorSettings = jsonObject.getJSONObject(FACE_AUTHENTICATION_SETTINGS)

            val filterStyle = if (faceAuthenticatorSettings.has(FILTER)) {
                CafFilterStyle.entries[faceAuthenticatorSettings.getInt(FILTER)]
            } else null

            CafFaceAuthenticationSettingsModel(
                faceAuthenticatorSettings.optBoolean(LOADING_SCREEN),
                faceAuthenticatorSettings.optBoolean(ENABLE_SCREEN_CAPTURE),
                filterStyle ?: CafFilterStyle.LINE_DRAWING
            )
        } else null
    }

    private companion object {
        const val STAGE = "stage"
        const val FACE_AUTHENTICATION_SETTINGS = "faceAuthenticationSettings"
        const val LOADING_SCREEN = "loadingScreen"
        const val ENABLE_SCREEN_CAPTURE = "enableScreenCapture"
        const val FILTER = "filter"
    }
}

File: caf-smart-auth-react-native-cli/ios/CafSmartAuthBridgeSettings.swift
import Foundation
import CafSmartAuth

internal struct CafFaceAuthenticationSettingsModel: Decodable {
  let loadingScreen: Bool?
  let filter: Int?
}

internal struct CafSmartAuthBridgeSettingsModel: Decodable {
  let stage: Int?
  let faceAuthenticationSettings: CafFaceAuthenticationSettingsModel?
}

internal class CafSmartAuthBridgeSettings {
  internal func parseJson(settings: String?) -> CafSmartAuthBridgeSettingsModel? {
    guard let data = settings?.data(using: .utf8) else {
      return nil
    }
    
    do {
      let decoder = JSONDecoder()
      let parsedSettings = try decoder.decode(CafSmartAuthBridgeSettingsModel.self, from: data)
      
      return parsedSettings
    } catch {
      return nil
    }
  }
}
